### ISSUE TITLE
Type detection fails in this circumstance

### ISSUE THREAD

**Issue by vilicvane** [microsoft/TypeScript#1165]
TypeScript 1.3 running on Visual Studio 2013 Update 3

Here's a minimum example that might trigger this bug. But it's really weird because sometimes it would work normally.

``` typescript
interface IMap<T> {
    [key: string]: T;
}

var map: IMap<string[]>;
var categories: string[];

each(categories, category => {
    var changes = map[category];

    return each(changes, change => {

    });
});

function each<T>(items: T[], handler: (item: T) => void) { }
```

The type of `changes` is detected as `any`, which should be `string[]`.

Changing `category =>` to `(category: string) =>` or removing `return` would fix. And sometimes other operations would also fix this.

It's also causing saving really slow BTW.
